File: frontend/src/lib/components/grid/messages.ts
import type { Position } from "./grid.svelte.ts";

interface LeadMsg {
	msg_type: 'set' | 'get' | 'error' | 'bulk' | 'eval';
	cell?: CellRef;
	raw?: string;
	eval?: Eval;
	eval_config?: EvalConfig;
	bulk_msgs?: Array<LeadMsg>;
}

interface CellRef {
	row: number;
	col: number;
}

type LiteralType = 'Number' | 'Boolean' | 'String';
type LiteralValue = number | string | boolean;
type EvalRange = Array<Eval>;

interface Literal {
	type: LiteralType;
	value: LiteralValue;
}

interface EvalCellRef {
	eval: Eval;
	reference: CellRef;
}

interface LeadErr {
	code: 'DivZero' | 'TypeErr' | 'Syntax' | 'Server' | 'Unsupported';
	desc: string;
	title: string;
}

interface EvalConfig {
	do_propagation: boolean;
	force_propagation: boolean;
}

// Tagged union
type Eval =
	| { literal: Literal }
	| { cellref: EvalCellRef }
	| { range: Range }
	| { err: LeadErr }
	| 'unset';

interface CellT {
	raw: string;
	temp_raw: string;
	pos: Position;
	temp_eval?: Eval;
	eval?: Eval;
}

export type { Eval, LeadMsg, LeadErr, Literal, CellRef, LiteralValue, CellT };

File: backend/src/messages.rs
use serde::{Deserialize, Serialize};

use crate::{cell::CellRef, evaluator::Eval};

#[derive(Serialize, Deserialize, Debug)]
#[serde(rename_all = "lowercase")]
pub enum MsgType {
    Set,
    Eval,
    Get,
    Error,
    Bulk,
}

#[derive(Serialize, Deserialize, Debug)]
pub struct EvalConfig {
    pub do_propagation: bool,
    pub force_propagation: bool,
}

impl Default for EvalConfig {
    fn default() -> Self {
        EvalConfig {
            do_propagation: true,
            force_propagation: false,
        }
    }
}

#[derive(Serialize, Deserialize, Debug)]
pub struct LeadMsg {
    pub msg_type: MsgType,
    pub cell: Option<CellRef>,
    pub raw: Option<String>,
    pub eval: Option<Eval>,
    pub eval_config: Option<EvalConfig>,
    pub bulk_msgs: Option<Vec<LeadMsg>>,
}
